### PR TITLE
Propagate logs to stdout when in k8s executor pod

### PR DESCRIPTION
**Pull request by dstandish**
Currently live logging (viewing from webserver while task running) for k8s executor pods doesn't work.

The cause has a few components from things we do during task run:
1. we remove the console handler from root logger.  So any logs that make it to root don't get emitted.
2. even if we keep console handler at root, it is designed to respect stdout redirection
3. even if we make it not respect stdout redirection in this context, task logger does not propagate, so anything going through task logger does not make it to root logger

The reason for (3) is we *copy* the task handler to root logger instead of *move*, which requires that we disable propagation at the task logger.

This PR fixes all of these things and, ultimately, simplifies the setup.

1. instead of keeping task handler in two places during execution -- airflow.task and root -- we keep it just in root.  This means we can let airflow.task logger propagate. this we now do always.
2. when running k8s executor pod, we keep the console handler at root and set it to not respect redirection, so that even though stdout gets redirected to task logger, when it makes it to root logger, it is still emitted to stdout.

### CODE REVIEW

File: airflow/cli/commands/task_command.py [hunk 1]
```diff
@@ -280,38 +281,80 @@ def _extract_external_executor_id(args) -> str | None:
...
+        root_level = root_logger.level
+        task_logger = ti.log
+        task_propagate = task_logger.propagate
+        task_handlers = task_logger.handlers.copy()
+
+        # if there are no task handlers, then we should not do anything
+        # because either the handlers were already moved by the LocalTaskJob
+        # invocation of task_run (which wraps the --raw invocation), or
+        # user is doing something custom / unexpected
+        if task_handlers:
```
Comment (uranusjr) [inline airflow/cli/commands/task_command.py]: I think it’d be good enough if we simply do a

```python
if not task_handlers:
    yield
    return
```

and get rid of the `did_modify` flag? This also allows everything after this to be unindented one level.

(I’m not even sure if the extra `yield` is necessary; probably?)

Comment (dstandish) [reply airflow/cli/commands/task_command.py]: yeah that should work thanks

Comment (dstandish) [reply airflow/cli/commands/task_command.py]: was just focused on getting this to work up to this point without worrying too much about quality :)

Comment (dstandish) [reply airflow/cli/commands/task_command.py]: removed the modify flag too

**Comment by dstandish** [comment]
two changes that are sort of elective in this one are (1) overridable "maintain propagate" and (2) checking if ti is actually a TI

both of these were required because while fixing some tests that were failing i noticed that the test replaced the whole logging config with something completely different and i curious why that was necessary so i dug in.  we needed to check TI because in some of the test it's a mock value (and checking doesn't hurt anything -- if it's not TI it will fail anyway.  then overriding maintain propagate is necessary to ensure that pytest caplog works... again... it doesn't hurt anything being there and allows us to do less manipulation in the tests which, all else equal, is a good thing.

**Comment by dstandish** [comment]
one thing i should add too though is....

with this change, we might actually be able to remove the maintain propagate logic entirely.  indeed we might be able to remove the recursive logger walking / propagate management in set_context as well.

the reason is, since we move FTH to root (rather than copy) propagation doesn't matter, because there's nothing you can propagate to beyond root.  it mattered when FTH was both at `airflow.task` and root because you could log twice.  

we wanted it at task because that's what `self.log` expected. and we wanted it at root so it could capture anything logged during task run (i.e. modules not children of airflow task).  but because we had it in both places, we had to be careful about propagation.  so this avoids that problem.

**Comment by malthe** [comment]
@dstandish I think an abstraction point in the executor framework would be better rather than the `if`-statements in this changeset. That is, we should make the executor pluggable in terms of supporting this logging capability.

In terms of actual implementation, rather than adjusting log handlers at runtime, perhaps an internal pub/sub mechanism could be devised.

**Comment by dstandish** [comment]
> @dstandish I think an abstraction point in the executor framework would be better rather than the `if`-statements in this changeset. That is, we should make the executor pluggable in terms of supporting this logging capability.
> 
> In terms of actual implementation, rather than adjusting log handlers at runtime, perhaps an internal pub/sub mechanism could be devised.
 
We already adjust handlers. This fixes a deficiency

File: airflow/cli/commands/task_command.py [hunk 2]
```diff
@@ -280,38 +281,85 @@ def _extract_external_executor_id(args) -> str | None:
...
     """
-    modify = not settings.DONOT_MODIFY_HANDLERS
-    if modify:
-        root_logger, task_logger = logging.getLogger(), logging.getLogger("airflow.task")
+    Redirect stdout to ti logger.
 
-        orig_level = root_logger.level
-        root_logger.setLevel(task_logger.level)
-        orig_handlers = root_logger.handlers.copy()
-        root_logger.handlers[:] = task_logger.handlers
```
Comment (dstandish) [inline airflow/cli/commands/task_command.py:304]: @malthe observe, we already copy (temporarily) to root logger the thandlers from airflow.task

This causes problems because we have to have complicated propagation rules at the airflow.task logger because we leave it there.

While my solution here is a few more lines (and a lot more comments) it's not very complicated.  There are two parts.  

1. Instead of _copying_ the handlers to root, I _move_ them to root.  We don't need them at task if they are already at root.  This could ultimately allow for simplification of our propagation logic.
2. Previously we _remove_ our console handler from root at run time.  Now, if we're in a k8s executor pod, I keep the console handler there.

**Comment by dstandish** [comment]
ok updated the description so hopefully it's a little clearer whats going on

**Comment by dstandish** [comment]
Ok @malthe i have simplified the code, if you would like to have another look:
https://github.com/apache/airflow/blob/0ade033eaa9feb52aecbe12d1b91131a27a80b4b/airflow/cli/commands/task_command.py#L322-L326

File: airflow/cli/commands/task_command.py [hunk 3]
```diff
@@ -281,38 +283,62 @@ def _extract_external_executor_id(args) -> str | None:
...
+        if not handler:
+            return
+        if handler not in logger.handlers:
+            logger.addHandler(handler)
+
+    # Move task handlers to root and reset task logger and restore original logger settings after exit.
+    # If k8s executor, we need to ensure that root logger has a console handler, so that
+    # task logs propagate to stdout (this is how webserver retrieves them while task is running).
+    root_logger = logging.getLogger()
+    console_handler = get_console_handler(root_logger)
```
Comment (uranusjr) [inline airflow/cli/commands/task_command.py]: Can be simplified into

```python
console_handler = next((h for h in logger.handlers if h.name == "console"), None)
```

Also it seems logic in `ensure_handler` can also be inlined?

File: airflow/utils/log/file_task_handler.py [hunk 4]
```diff
@@ -62,19 +62,29 @@ def __init__(self, base_log_folder: str, filename_template: str | None = None):
...
-        self.handler.setLevel(self.level)
-        return None
+        from airflow.models.taskinstance import TaskInstance
+
+        if isinstance(ti, TaskInstance):
+            local_loc = self._init_file(ti)
+            self.handler = NonCachingFileHandler(local_loc, encoding="utf-8")
+            if self.formatter:
+                self.handler.setFormatter(self.formatter)
+            self.handler.setLevel(self.level)
```
Comment (uranusjr) [inline airflow/utils/log/file_task_handler.py]: When is this not true…?

File: airflow/cli/commands/task_command.py [hunk 5]
```diff
@@ -281,38 +283,62 @@ def _extract_external_executor_id(args) -> str | None:
...
+    with LoggerMutationHelper(root_logger), LoggerMutationHelper(ti.log) as task_helper:
+        task_helper.move(root_logger)
+        if IS_K8S_EXECUTOR_POD:
```
Comment (malthe) [inline airflow/cli/commands/task_command.py:308]: I think we ought to have enough abstraction in the executor interface that we don't need this `if`-clause here.

File: airflow/utils/log/file_task_handler.py [hunk 6]
```diff
@@ -62,19 +62,29 @@ def __init__(self, base_log_folder: str, filename_template: str | None = None):
                 # handler, not the one that calls super()__init__.
                 stacklevel=(2 if type(self) == FileTaskHandler else 3),
             )
+        self.maintain_propagate: bool = False
 
     def set_context(self, ti: TaskInstance) -> None | SetContextPropagate:
         """
         Provide task_instance context to airflow task handler.
 
+        Generally speaking returns None.  But if attr `maintain_propagate` has
```
Comment (malthe) [inline airflow/utils/log/file_task_handler.py:76]: It might be possible to use mocking instead to avoid having functionality that's used only in unit testing.

File: airflow/utils/log/file_task_handler.py [hunk 7]
```diff
@@ -92,16 +102,17 @@ def _render_filename(self, ti: TaskInstance, try_number: int) -> str:
         with create_session() as session:
             dag_run = ti.get_dagrun(session=session)
             template = dag_run.get_log_template(session=session).filename
-        str_tpl, jinja_tpl = parse_template_string(template)
+            str_tpl, jinja_tpl = parse_template_string(template)
```
Comment (malthe) [inline airflow/utils/log/file_task_handler.py:107]: Why is this parsed within the session context now?

Comment (dstandish) [reply airflow/utils/log/file_task_handler.py:107]: if you look a few lines below, we need to keep `context = ti.get_template_context(session=session)`
in the same session.  otherwise it creates a new session (provided by provide_session), and this can cause problems post-os.fork.  this is why the test for standard task runner previously blew away the entire log config because. with this, we don't have to do that anymore.

File: airflow/cli/commands/task_command.py (hunk 2 above)
Comment (malthe) [reply airflow/cli/commands/task_command.py:304]: I didn't quite understand what "stdout redirection" means – is that within Python or is that something happening outside of Python (in a shell script perhaps) – ?

When I think about this problem, it seems that there is an orthogonal concern in the logging setup of Airflow which is whether or not the task logger (during execution) should be emitted to the stdout stream.

That's something you could want in any situation, not just K8S.

File: airflow/cli/commands/task_command.py (hunk 5 above)
Comment (dstandish) [reply airflow/cli/commands/task_command.py:308]: a noble goal. but i would say out of scope for this PR.  there already is this env var.  and if executor provides an interface we can use here, we can later update to use it.

File: airflow/cli/commands/task_command.py (hunk 2 above)
Comment (dstandish) [reply airflow/cli/commands/task_command.py:304]: > I didn't quite understand what "stdout redirection" means – is that within Python or is that something happening outside of Python (in a shell script perhaps) – ?

i am not sure what comment you refer to but, what this means is there are stdlib helpers to take things that would ordinarily go to stdout (e.g. print) and send them to some other stream. 

see `from contextlib import redirect_stdout`.  it just monkey patches `sys.stdout` temporarily.

we use them (for better or worse) to redirect stdout to task logs.

our log read logic when task is running is usually, read from flask log server.  most task logging is redirected to log (and therefore file) and our "console" handler respects this redirection.   for obvious reasons -- the celery worker log, or local executor log would get unreasonably chatty.  but in k8s executor context, there is no log server on the worker, and there is no problem with keeping the stdout, and importantly our log read logic _assumes_ everything will be forwarded to stdout -- and this is what's broken that i am here fixing.

File: airflow/utils/log/file_task_handler.py (hunk 6 above)
Comment (dstandish) [reply airflow/utils/log/file_task_handler.py:76]: in this case, to do so, we would have to obliterate code that i'd rather keep. if we mack set_context (or the free function of same name) then we don't execute them.  doing this lets us keep that code live.  it ends up removed in this followup PR (https://github.com/apache/airflow/pull/28571)

which i need to rebase here after some recent changes... and the diff will be horrendous for now don't bother looking :)

Comment (dstandish) [reply airflow/utils/log/file_task_handler.py:76]: i've updated it to mark it as `:meta private:` to indicate should not be used by users / not part of public interface / can be removed at any time.

File: airflow/utils/log/file_task_handler.py (hunk 4 above)
Comment (dstandish) [reply airflow/utils/log/file_task_handler.py]: i think it came up when i was mocking a TI.... and  it doesn't hurt anybody.... but i've removed it for now let's see what the tests say

File: airflow/cli/commands/task_command.py (hunk 3 above)
Comment (dstandish) [reply airflow/cli/commands/task_command.py]: done

File: airflow/utils/log/file_task_handler.py (hunk 4 above)
Comment (malthe) [reply airflow/utils/log/file_task_handler.py]: It does hurt somebody if they end up scratching their head to make sense of it!

File: airflow/utils/log/file_task_handler.py (hunk 6 above)
Comment (malthe) [reply airflow/utils/log/file_task_handler.py:76]: I don't really get this one. The `self.maintain_propagate` attribute is available for anyone to check as it is. The point of the method is to set the handler to refer to the provided `ti`.

Comment (dstandish) [reply airflow/utils/log/file_task_handler.py:76]: Look at how it is used in the standard task runner test

Comment (dstandish) [reply airflow/utils/log/file_task_handler.py:76]: Meta private affects documentation and has been used to signal not public api

Comment (dstandish) [reply airflow/utils/log/file_task_handler.py:76]: Not mocking and not replacing entire logging config with a fake one is how I discovered we were creating more connections unnecessarily

File: airflow/utils/log/file_task_handler.py (hunk 4 above)
Comment (dstandish) [reply airflow/utils/log/file_task_handler.py]: well... i guess there are degrees of hurt...

anyway, it was needed because something in existing test was mocking a TI and the actual methods don't like the mock obj

and since i restored normal logging config, now the mocked ti is hitting handler. anyway, mocked this method away so non issue now.  removed the if, and tests should go green again.

File: airflow/cli/commands/task_command.py [hunk 8]
```diff
@@ -390,11 +400,19 @@ def task_run(args, dag=None):
 
     log.info("Running %s on host %s", ti, hostname)
 
+    # IMPORTANT, have to re-configure ORM with the NullPool, otherwise, each "run" command may leave
+    # behind multiple open sleeping connections while heartbeating, which could
+    # easily exceed the database connection limit when
+    # processing hundreds of simultaneous tasks.
+    # this should be last thing before running, to reduce likelihood of an open session
+    # which can cause trouble if running process in a fork.
+    settings.reconfigure_orm(disable_connection_pool=True)
```
Comment (ashb) [inline airflow/cli/commands/task_command.py:409]: Is this related to this logging PR in anyway?

File: airflow/utils/log/file_task_handler.py [hunk 9]
```diff
@@ -62,19 +62,29 @@ def __init__(self, base_log_folder: str, filename_template: str | None = None):
                 # handler, not the one that calls super()__init__.
                 stacklevel=(2 if type(self) == FileTaskHandler else 3),
             )
+        self.maintain_propagate: bool = False
+        """If true, overrides default behavior of setting propagate=False
```
Comment (ashb) [inline airflow/utils/log/file_task_handler.py]:
```suggestion
        """
        If true, overrides default behavior of setting propagate=False
        
```

Comment (ashb) [reply airflow/utils/log/file_task_handler.py]: (The blank line is the important bit here)

File: tests/cli/commands/test_task_command.py [hunk 10]
```diff
@@ -69,6 +72,15 @@ def move_back(old_path, new_path):
     shutil.move(new_path, old_path)
 
 
+@pytest.fixture()
+def freeze_time():
+    timestamp = "2022-06-10T12:02:44+00:00"
+    freezer = time_machine.travel(timestamp, tick=False)
+    freezer.start()
+    yield
+    freezer.stop()
```
Comment (ashb) [inline tests/cli/commands/test_task_command.py]: This fixture doesn't seem to be used anywhere?

File: tests/cli/commands/test_task_command.py [hunk 11]
```diff
@@ -600,6 +612,48 @@ def test_external_executor_id_present_for_process_run_task(self, mock_local_job)
                 external_executor_id="ABCD12345",
             )
 
+    @pytest.mark.parametrize("is_k8s", ["true", ""])
+    def test_logging_with_run_task_stdout_k8s_executor_pod(self, is_k8s):
+        """
+        When running task --local as k8l executor pod, all logging should make it to stdout.
```
Comment (ashb) [inline tests/cli/commands/test_task_command.py]:
```suggestion
        When running task --local as k8s executor pod, all logging should make it to stdout.
```

File: tests/task/task_runner/test_standard_task_runner.py [hunk 12]
```diff
@@ -131,38 +141,34 @@ def test_notifies_about_start_and_stop(self):
...
+            data_interval=(DEFAULT_DATE, DEFAULT_DATE),
+            state=State.RUNNING,
+            start_date=DEFAULT_DATE,
+        )
+        ti = TaskInstance(task=task, run_id="test")
+        job1 = LocalTaskJob(task_instance=ti, ignore_ti_state=True)
+        runner = StandardTaskRunner(job1)
+        runner.start()
+
+        # Wait until process sets its pgid to be equal to pid
```
Comment (ashb) [inline tests/task/task_runner/test_standard_task_runner.py]:
```suggestion
        # Wait until process makes itself the leader of it's own process group
```

Comment (ashb) [reply tests/task/task_runner/test_standard_task_runner.py]: Though why do we have this loop, and then a 10s timeout directly below it?

**Review by ashb — APPROVED** [review]
A few small nits, but looks good to me overall.

File: tests/cli/commands/test_task_command.py (hunk 10 above)
Comment (dstandish) [reply tests/cli/commands/test_task_command.py]: yup, removed

File: tests/task/task_runner/test_standard_task_runner.py (hunk 12 above)
Comment (dstandish) [reply tests/task/task_runner/test_standard_task_runner.py]: do not know, existing code :)

Comment (dstandish) [reply tests/task/task_runner/test_standard_task_runner.py]: looking, though, i think the timeout is related to "done" while the loop is about "started"

File: airflow/cli/commands/task_command.py (hunk 8 above)
Comment (dstandish) [reply airflow/cli/commands/task_command.py:409]: i don't *think* this will fail if we remove this change.  you could call it a drive by and i can remove if you like.  

task runner tests failed with this change. then i was like "oh why is it just throwing away whole log config, let's try and not do that". then after leaving the FTH intact for that test, that i discovered there was an extraneous session being created in created in _render_filename.  as part of troubleshooting the failures and discovering that though, also saw that this reconfigure is happening earlier than it probably ought to. should be done as late as  possible i think.
